src/ray/gcs/redis_context.cc

```diff
@@ -656,27 +676,55 @@ std::unique_ptr<CallbackReply> RedisContext::RunArgvSync(
     argv.push_back(arg.data());
     argc.push_back(arg.size());
   }
-  auto redis_reply = reinterpret_cast<redisReply *>(
-      ::redisCommandArgv(context_.get(), args.size(), argv.data(), argc.data()));
-  if (redis_reply == nullptr) {
-    RAY_LOG(ERROR) << "Failed to send redis command (sync): " << context_->errstr;
-    return nullptr;
+  // Run the command. We try to reconnect if the connection is lost, and we retry with
+  // exponential backoff using the same connection if error happens.
+  auto exp_back_off = ExponentialBackOff(RayConfig::instance().redis_retry_base_ms(),
+                                         RayConfig::instance().redis_retry_multiplier(),
+                                         RayConfig::instance().redis_retry_max_ms());
+  size_t pending_retries = RayConfig::instance().num_redis_request_retries() + 1;
+  std::unique_ptr<redisReply, decltype(freeReplyObject) *> redis_reply(nullptr,
+                                                                       freeReplyObject);
+  while (pending_retries > 0) {
+    redis_reply.reset(reinterpret_cast<redisReply *>(
+        ::redisCommandArgv(context_.get(), args.size(), argv.data(), argc.data())));
+    // Disconnected. Try to reconnect first.
+    if (redis_reply == nullptr) {
+      RAY_CHECK_OK(this->Reconnect());
+      continue;
+    }
+    // Error happened, retry with same connection.
+    if (redis_reply->type != REDIS_REPLY_ERROR) {
+      std::unique_ptr<CallbackReply> callback_reply(new CallbackReply(redis_reply.get()));
```

Comment:
Prefer `std::make_unique`.
```
return std::make_unique<CbRly>(reply);
```
Ref: https://stackoverflow.com/questions/22571202/differences-between-stdmake-unique-and-stdunique-ptr-with-new

Reply:
My personal preferrance: update `CallbackReply`'s ctor to take a constant reference.
Reason:
- pointer needs to check nullability, reference doesn't
- non-const pointer implies mutability and potential memory ownership transfer, while here we're simply parsing the reply